### ISSUE TITLE
Investigate uncrustify

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#6344]
Investigate [uncrustify](https://github.com/uncrustify/uncrustify) as a reformatting tool for https://github.com/Mbed-TLS/mbedtls/issues/6343.

We've investigated clang-format, see https://github.com/Mbed-TLS/mbedtls/issues/4887. It has some limitations. Can uncrustify help? Does uncrustify have different limitations that would annoy us?

Some clang-format limitations:

* Only modifies whitespace, can't e.g. remove parentheses around return values, or add/remove braces.
* Breaks alignment in places where it's nice to have (e.g. series of `#define NAME value` with aligned values), and apparently can't be told not to.
* Difficult to build if you can't get binaries for the right version on your platform (and the output isn't guaranteed stable across versions, so users need the same version as us). Maybe not important since prebuilt clang binaries are available for most platforms (but may be blocked by IT policies; uncrustify is easy to build from source).

**Comment by gilles-peskine-arm** [comment]
> Pros:

Looks pretty nice!

> Cons:

> Not always idempotent - needs care when configuring to ensure it is.

That's an issue, but as long as we find a working configuration, we're good. We should remember that it's an issue and check this when we do the transformations, in case we don't immediately discover the non-idempotency cases in our configuration.

> Currently doesn't deal with braces spanning preprocessor definitions (stated in the known issues) which may cause problems in our codebase.

We used to have this kind of problematic cases, but they were also hard to read for humans and I would reject them in new code. If this is a motivation for getting rid of any such remaining usage, I see it as more of a pro.

> Occasionally generates invalid output, for example here.

That's the same kind of problem as the previous one. Evidently there's no consensus on the team that this kind of code is problematic.

> Doesn't deal well with inventive macros like [this one](https://github.com/davidhorstmann-arm/mbedtls/blob/uncrustify-experimentation/library/aria.c#L892) in aria.c that expand to partial statements.

Invalid output is a problem. But I would have rejected this particular code in review, and once again I see it as a pro that we'd be fixing it. So here the question is, are we confident that we can detect the problematic cases?

**Comment by gilles-peskine-arm** [comment]
Having progressed in our study of uncrustify, we've decided to go for it. Known limitations:

* Different versions have different bugs and features. So we'll use one specific version.
* Uncrustify can mess up in edge cases. But if the code passes the tests, we're reasonably confident. And we will of course do significant spot checks.
* It seems we'll have to run uncrustify twice to make it idempotent.
